### PR TITLE
V1 embeddings

### PR DESCRIPTION
**Pull request by maxdebayser**
# Description

This PR enables embedding models on vllm V1. In contrast with the V1 GPU implementation, here I added a separate model runner because for most of the embedding models there is no need for continuous batching. To avoid code repetition, I refactored the input batch and model runner classes into a class hierarchy with common base classes.

@gmarinho2 contributed a test that verifies that the returned embeddings don't change with batch size.

### CODE REVIEW

File: vllm_spyre/v1/worker/spyre_base_input_batch.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+_SAMPLING_EPS = 1e-5
+
+
+@dataclass
+class BaseRequestState:
+
+    req_id: str
+    prompt_token_ids: list[int]
+    left_padding: int = 0  # Defaults to 0, i. e. not padding
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_base_input_batch.py:21]: this is only used in `SamplingRequestState`, right? could be moved to the derived class.

File: tests/v1/worker/test_spyre_input_batch.py [hunk 2]
```diff
@@ -11,17 +11,17 @@
                                              init_builtin_logitsprocs)
 from vllm.v1.sample.metadata import SamplingMetadata
 
-from vllm_spyre.v1.worker.spyre_input_batch import (CachedRequestState,
-                                                    InputBatch)
+from vllm_spyre.v1.worker.spyre_input_batch import (SamplingInputBatch,
+                                                    SamplingRequestState)
```
[the remark this answers is not in the document]
Comment (yannicks1) [reply tests/v1/worker/test_spyre_input_batch.py:15]: having read more of the PR, I now understand 😄

File: vllm_spyre/v1/worker/spyre_base_input_batch.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        # Copy the prompt token ids.
+        num_prompt_tokens = len(request.prompt_token_ids)
+        self.num_prompt_tokens[req_index] = num_prompt_tokens
+
+        self.token_ids_cpu[
+            req_index, :num_prompt_tokens] = request.prompt_token_ids
+
+        self._num_requests += 1
+        assert self._num_requests <= self.max_num_reqs
+        return req_index
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_base_input_batch.py:112]: a bit strange to have `add_request` alongside `_add_request`. 
you have a `BaseInputBatch.add_request` and a `SamplingInputBatch.add_request` in which you call `super().add_request` (same for `PoolingInputBatch`), that works...

File: vllm_spyre/v1/worker/spyre_base_input_batch.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        self.req_id_to_index = {}
+
+        self._req_ids = [None] * self.max_num_reqs
+
+        self._num_requests = 0
+
+    def remove_request(self, req_id: str):
+        raise NotImplementedError
+
+    def _remove_request(self, req_id: str) -> Optional[int]:
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_base_input_batch.py:127]: same as above, you don't need a separate `_remove_request` function...

File: vllm_spyre/v1/worker/spyre_base_input_batch.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+        # Remove the references
+
+        self._req_ids[req_index] = None
+        self._num_requests -= 1
+        return req_index
+
+    def refresh_metadata(self):
+        pass
+
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_base_input_batch.py:154]: can that be removed when removing the line 164 and line 231 in `spyre_pooling_model_runner.py`. It looks to me like it might be doing nothing.

**Review by yannicks1 — COMMENTED** [review]
I am through 7/10 files, need to continue later... This is a huge PR:)

File: vllm_spyre/v1/worker/spyre_pooling_model_runner.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+import vllm_spyre.envs as envs_spyre
+
+logger = init_logger(__name__)
+
+BACKEND_LIST = ['sendnn', 'inductor']
+
+
+@dataclass(frozen=True)
+class PoolingForwardInputs(ModelForwardInputs):
+
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_pooling_model_runner.py:41]: analogous to `SamplingForwardInputs` maybe add:
```
"""
Used by the SpyrePoolingModelRunner.
"""
```

File: vllm_spyre/v1/worker/spyre_pooling_model_runner.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        input_ids = torch.stack(padded_input_ids_list)
+        if padded_token_type_list:
+            token_type_ids = torch.stack(padded_token_type_list)
+        else:
+            token_type_ids = None
+        mask = torch.stack(mask_list)
+        position_ids = torch.stack(position_ids_list)
+
+        return input_ids, token_type_ids, position_ids, mask
+
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_pooling_model_runner.py:114]: if you can move `update_states()` to here, then it is consistent with the order in the other Model runners:)

File: vllm_spyre/v1/worker/spyre_model_runner.py [hunk 8]
```diff
@@ -441,13 +521,53 @@ def execute_model(
...
             prompt_logprobs_dict=prompt_logprobs_dicts,
-            pooler_output=None,
+            pooler_output=[],
         )
 
         return model_output
 
 
-class StaticBatchingSpyreModelRunner(SpyreModelRunner):
+class WarmupShapesMixin:
```
Comment (yannicks1) [inline vllm_spyre/v1/worker/spyre_model_runner.py:530]: hmm, might there be another, more elegant way to distribute the method `_get_padded_batch_size` to both `SpyrePoolingModelRunner`, and `StaticBatchingSpyreModelRunner` ? Also, the name `WarmupShapesMixin` sounds a bit weird to me 😄 . Tagging our most experience SWE here:) @joerunde @tjohnson31415

File: tests/v1/worker/test_spyre_input_batch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxdebayser) [reply tests/v1/worker/test_spyre_input_batch.py:15]: Yeah, just to make clear for others who might read this later, the `CachedRequestState` class was refactored into a base class `BaseRequestState` that holds common attributes and methods and two derived classes `SamplingRequestState` and `PoolingRequestState` that are specialized for the generation and pooling use cases.

File: vllm_spyre/platform.py [hunk 9]
```diff
@@ -243,7 +271,7 @@ def supports_v1(cls, model_config: ModelConfig) -> bool:
         model configuration.
         """
         # We don't have an embedding runner for v1 yet
```
[the remark this answers is not in the document]
Comment (maxdebayser) [reply vllm_spyre/platform.py:273]: Yes, thanks, for pointing this out!

File: vllm_spyre/v1/worker/spyre_base_input_batch.py (hunk 5 above)
Comment (maxdebayser) [reply vllm_spyre/v1/worker/spyre_base_input_batch.py:154]: Yes, good point. It looks like an artifact of replicating the structure of the upstream model runner that handles both pooling and generation.

File: vllm_spyre/v1/worker/spyre_base_input_batch.py (hunk 3 above)
Comment (maxdebayser) [reply vllm_spyre/v1/worker/spyre_base_input_batch.py:112]: Oh yeah, this is what you would have to do preserve Liskov's substitution principle, but since we have python generics it's not required.

File: tests/e2e/test_spyre_embeddings.py [hunk 10]
```diff
@@ -49,3 +53,59 @@ def test_output(
...
+    reference_embeds = st_embeddings(model, prompts)
+
+    vllm_model = LLM(model=model,
+                     task="embed",
+                     tokenizer=model,
+                     max_model_len=256,
+                     block_size=256,
+                     tensor_parallel_size=1)
+
+    def chunk_embeds(step):
```
Comment (sducouedic) [inline tests/e2e/test_spyre_embeddings.py:86]: Wouldn't it make more sense to name this function `batch_embeds` instead of `chunk_embeds`? given that this function embed 1 prompt, then 2 prompts, then 4 prompts.

File: tests/e2e/test_spyre_embeddings.py [hunk 11]
```diff
@@ -49,3 +53,59 @@ def test_output(
...
+    model,
+    backend,
+    warmup_shapes,
+    vllm_version,
+) -> None:
+
+    os.environ["VLLM_SPYRE_DYNAMO_BACKEND"] = backend
+    os.environ['VLLM_USE_V1'] = "1" if vllm_version == "V1" else "0"
+
+    prompts = get_chicken_soup_prompts(4)
```
Comment (sducouedic) [inline tests/e2e/test_spyre_embeddings.py:76]: I think some chicken prompt soup won't fit the warmup shape of 64

File: tests/e2e/test_spyre_embeddings.py [hunk 12]
```diff
@@ -49,3 +53,59 @@ def test_output(
...
+    (64, 2),
+    (64, 4),
+])  # (prompt_length/batch_size)
+@pytest.mark.parametrize("backend", get_spyre_backend_list())
+@pytest.mark.parametrize("model", get_spyre_model_list(isEmbeddings=True))
+@pytest.mark.parametrize("vllm_version", ["V0", "V1"])
+def test_scheduling_invariance(
+    model,
+    backend,
+    warmup_shapes,
```
Comment (sducouedic) [inline tests/e2e/test_spyre_embeddings.py:69]: I don't see where the shape is actually exported to the environment variables? ie. `VLLM_SPYRE_WARMUP_PROMPT_LENS` and `VLLM_SPYRE_WARMUP_BATCH_SIZES`

File: tests/e2e/test_spyre_embeddings.py [hunk 13]
```diff
@@ -49,3 +53,59 @@ def test_output(
...
+                     tokenizer=model,
+                     max_model_len=256,
+                     block_size=256,
+                     tensor_parallel_size=1)
+
+    def chunk_embeds(step):
+        vllm_outputs = []
+        for i in range(0, len(prompts), step):
+            emb_outputs = [
+                req.outputs for req in vllm_model.embed(prompts[i:i + step])
```
Comment (sducouedic) [inline tests/e2e/test_spyre_embeddings.py:90]:
```suggestion
                req.outputs.embedding for req in vllm_model.embed(prompts[i:i + step])
```

File: tests/e2e/test_spyre_embeddings.py (hunk 12 above)
Comment (maxdebayser) [reply tests/e2e/test_spyre_embeddings.py]: Yes, it wasn't :D

**Comment by maxdebayser** [comment]
All tests are passing now after the changes from the first round of reviews.

File: vllm_spyre/v1/worker/spyre_model_runner.py [hunk 14]
```diff
@@ -37,6 +37,8 @@
 
 from vllm.v1.outputs import EMPTY_MODEL_RUNNER_OUTPUT, ModelRunnerOutput
 
+# TODO: fix import when we upgrade to vLLM 0.9.3
```
Comment (prashantgupta24) [inline vllm_spyre/v1/worker/spyre_model_runner.py:40]: Very minor nit but vllm released 0.10.0 instead :)

File: vllm_spyre/v1/worker/spyre_pooling_input_batch.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+                self.token_ids_cpu_tensor.shape,
+                device="cpu",
+                dtype=torch.int8,
+                pin_memory=False,
+            )
+            self._token_type_ids_cpu = cast(
+                torch.Tensor, self.token_type_ids_cpu_tensor).numpy()
+        return self._token_type_ids_cpu
+
+    def has_token_types(self) -> bool:
```
Comment (wallashss) [inline vllm_spyre/v1/worker/spyre_pooling_input_batch.py:66]: I'm pretty sure that this method is not being used and, I was wondering if the tensors related to that like token_type_ids_cpu_tensor are being used as well. I mean, I could find it being populated, but not being read if I did the reading correctly.

Comment (maxdebayser) [reply vllm_spyre/v1/worker/spyre_pooling_input_batch.py:66]: Yes, this is to prepare for changes that haven't been merged upstream yet. I can remove these to simplify the PR for now.

**Review by wallashss — APPROVED** [review]
LGTM! Thanks for the refactoring of input batch, it makes sense to me. 

Just a friendly reminder: update `docs/supported_features.md`
